godot-ffi/src/binding/multi_threaded.rs

```diff
@@ -0,0 +1,129 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+//! Thread safe binding storage.
+//!
+//! This can be used from different threads without issue, as late initialization uses `OnceLock`.
+//!
+//! The user of these structs and functions must still ensure that multi-threaded usage of the various pointers is safe.
+
+use std::sync::OnceLock;
+
+use crate::{
+    BuiltinLifecycleTable, BuiltinMethodTable, ClassEditorMethodTable, ClassSceneMethodTable,
+    ClassServersMethodTable, GDExtensionInterface, GdextRuntimeMetadata, UtilityFunctionTable,
+};
+
+use super::GDExtensionClassLibraryPtr;
+
+pub(super) struct BindingStorage {
+    binding: OnceLock<GodotBinding>,
+}
+
+impl BindingStorage {
+    /// Get the static binding storage.
+    #[inline(always)]
+    fn storage() -> &'static Self {
+        static BINDING: BindingStorage = BindingStorage {
+            binding: OnceLock::new(),
+        };
+
+        &BINDING
+    }
+
+    /// Initialize the binding storage, this must be called before any other public functions.
+    #[must_use]
+    pub fn initialize(binding: GodotBinding) -> Option<()> {
+        let storage = Self::storage();
+
+        storage.binding.set(binding).ok()?;
+
+        Some(())
+    }
+
+    /// Get the binding from the binding storage.
+    ///
+    /// # Safety
+    /// - The binding must be initialized.
+    #[inline(always)]
+    pub unsafe fn get_binding_unchecked() -> &'static GodotBinding {
+        let storage = Self::storage();
+        let binding = storage.binding.get();
+
+        debug_assert!(
+            binding.is_some(),
+            "Godot engine not available; make sure you are not calling it from unit/doc tests"
+        );
+
+        // SAFETY: `binding` is `None` when the binding is uninitialized, but the safety invariant of this method is that
+        // the binding is initialized.
+        unsafe { binding.unwrap_unchecked() }
+    }
+
+    pub fn is_initialized() -> bool {
+        let storage = Self::storage();
+        storage.binding.get().is_some()
+    }
+}
+
+// Note, this is `Sync` and `Send` because all its fields are. We have avoided implementing `Sync` and `Send` for `GodotBinding`
+// as that could hide issues if any of the field types are changed to no longer be sync/send, but the manual implementation for
+// `GodotBinding` wouldn't detect that.
+pub struct GodotBinding {
+    pub(super) interface: GDExtensionInterface,
+    pub(super) library: GDExtensionClassLibraryPtr,
+    pub(super) global_method_table: BuiltinLifecycleTable,
+    pub(super) class_server_method_table: OnceLock<ClassServersMethodTable>,
+    pub(super) class_scene_method_table: OnceLock<ClassSceneMethodTable>,
+    pub(super) class_editor_method_table: OnceLock<ClassEditorMethodTable>,
+    pub(super) builtin_method_table: OnceLock<BuiltinMethodTable>,
```

Comment:
Also here, let's avoid double locking indirection.

Reply:
Here too i dont think there's really much value to doing `foo.as_ref().unwrap_unchecked()` as opposed to `foo.get().unwrap_unchecked()`.

Reply:
But here it's different -- Once*Lock* has thread synchronization overhead that is simply never needed.

Even if it achieves the same performance through `unwrap_unchecked`, I'd argue that `OnceLock` expresses the intent _worse_ than `Option` or `OnceCell` because there is no synchronization involved -- the outer lock already handles that.

Maybe we can use `OnceCell` here instead -- would this allow to share the code with the single-threaded implementation? If yes, that alone is reason to change in my opinion.

Reply:
Is your argument that because the struct is already in a `OnceLock`, these inner ones are unnecessary because access is already synchronized? Because if it is then that's wrong. `OnceLock` does not synchronize access, it only synchronizes initialization. Any future access after initialization is unsynchronized.

That's for instance why `OnceLock<T>` is only `Sync` if `T` is `Sync` (and `Send` actually).

Reply:
You're right, however my points still stand:
1. We don't need synchronization even at initialization.
2. `OnceCell` or `Option` are simpler and can do the job equally well.
3. Using those would allow do un-duplicate the code between the two implementations.

Reply:
> > `OnceLock` does not synchronize access, it only synchronizes initialization. Any future access after initialization is unsynchronized.
> 
> How does that work? Access and initialization are mixed in the `get_or_init()` method. Even `get()` must check whether initialization has already run -- and since `get()` can be called from multiple threads, there needs to be synchronization. The init flag (a `Once` object) is synchronized.

Yes, the init flag is synchronized. So access relative to the initialization is synchronized. But accesses after initialization are not. I.e two different threads can race wrt. writing to the struct stored in the `OnceLock` after it has been initialized. 

>     1. We don't need synchronization even at initialization -- the init levels are called from one thread.

If that is a guarantee from godot that we can rely on then sure. 

>     2. `OnceCell` or `Option` are simpler and can do the job equally well.

As everything works currently, neither `Option` nor `OnceCell` do the job equally well. At least without other modifications such as newtype wrapping.

When the binding is initialized, we initialize a `OnceLock<GodotBinding>` in a `static`. Crucially this means two things:
1. `GodotBinding` must be `Sync`, and preferably we dont just implement `Sync` for `GodotBinding`
2. any access to `GodotBinding` after initialization, must go through a `&GodotBinding` reference

The first point rules out `OnceCell` for the fields, since `OnceCell` is not `Sync`. We could newtype wrap `OnceCell` but that seems undesirable to me.

The second point rules out `Option`, since once the field is set to `None` and the static has been initialized, it will be impossible to modify the field to be a `Some`. That is because we can only get a `&` reference to the field, and it would be UB to modify the `Option` through that. So then we would actually need a `UnsafeCell<Option<T>>`, however `UnsafeCell` is not `Sync`, which brings us back to the issue that `OnceCell` has.

>     3. Using those would allow do un-duplicate the code between the two implementations.

Sure, so would using `OnceLock` in both. In fact that would be the simplest way to unduplicate both and from what i can tell it probably wont have a performance impact. But also at that point we might as well just use `OnceLock` everywhere and merge the two implementations like that.

Reply:
I think the options we have are basically:

1. Code duplication
2. Use `OnceLock` everywhere 
3. Something like `UnsafeOnceCell`
4. Fudging with newtypes to use `OnceCell` everywhere
5. Leave `static mut` in place, plan to switch to `SyncUnsafeCell`

I'd be fine with any of 1-3, i'd rather avoid 4 cause it'd be rather messy and feels like just a worse version of 3. My main issue with 5 is that it's very easy to accidentally mess up and break things without realizing it, since both `static mut` and `SyncUnsafeCell` leaves a lot of stuff up to the user.

Reply:
> Yes, that's why i use `unwrap_unchecked`, since the only way it could fail is if the `OnceLock` is uninitialized

No, but you don't -- `OnceLock` has no `unwrap_unchecked()`.
You call `OnceLock::get()` followed by `Option::unwrap_unchecked()`.

`get()` is not free! 🙂 https://doc.rust-lang.org/stable/src/std/sync/once_lock.rs.html#78-85
```rs
    pub fn get(&self) -> Option<&T> {
        if self.is_initialized() {
            // Safe b/c checked is_initialized
            Some(unsafe { self.get_unchecked() })
        } else {
            None
        }
    }

    fn is_initialized(&self) -> bool {
        self.once.is_completed() // <-- synchronization here!
    }
```

---

> I mean i can just make a `UnsafeOnceCell` which is just an `UnsafeCell<Option<T>>`. And just use that instead of `OnceCell/OnceLock` everywhere. Should simplify a lot of things. 

So you want to add _another_ abstraction on top? 😁

No, I think the design you proposed could work. The more locks/cells it can replace, the better. Let's just please keep this as lightweight as possible. The `static mut` solution may be frowned upon, but it would only need a bit of tweaking for the multithreaded case to be functional and sound -- and **this** should be the bar, not some random idiomatic Rust. Sometimes abstractions are very useful (`godot-cell`), but here we "just" need global initialization. So let's always focus on the wider context in which we're solving a particular problem 🙂 

We may have to rethink the outer lock as well. The same logic applies there: most of the inner data structures are guaranteed to be readable race-free once the init phase is over, so the outer lock is also unnecessary. There are some cases where we need to be careful with lazy initialization (some configs + method tables in `lazy-function-tables` feature). But the Godot APIs should mostly be readable without going through locks, even in multi-threaded code.*

Whether Godot APIs are _callable_ in a thread-safe way is of course another question. However that cannot easily be addressed by a global mutex alone, because we'd need to prevent handing out any function pointer. I guess that's topic for another day, but generally there seems to be lots of potential for accessing global data in parallel once the init phase is over.

---

Regarding Rust, I'm a bit disappointed that its tools clearly fail to deliver here. Apparently, in almost a decade since 1.0, no one ever encountered the scenario "first init, then everyone reads" in the real world, where the "everyone reads" part should be free. At least not enough people to standardize it. `SyncUnsafeCell` is a start, but it's quite generic and from its motivation seems more a compromise from people who have a categoric issue with `static mut`.

This is a recurring issue in Rust. Having seen the C++ side, it's good to have safety, but it's bad to design _only_ around it.

Reply:
> > Yes, that's why i use `unwrap_unchecked`, since the only way it could fail is if the `OnceLock` is uninitialized
> 
> No, but you don't -- `OnceLock` has no `unwrap_unchecked()`. You call `OnceLock::get()` followed by `Option::unwrap_unchecked()`.
> 
> `get()` is not free! 🙂 https://doc.rust-lang.org/stable/src/std/sync/once_lock.rs.html#78-85 and https://github.com/rust-lang/rust/blob/11f32b73e0dc9287e305b5b9980d24aecdc8c17f/library/std/src/sys_common/once/futex.rs#L75-L80
> 
> ```rust
> impl OnceLock {
>     pub fn get(&self) -> Option<&T> {
>         if self.is_initialized() {
>             // Safe b/c checked is_initialized
>             Some(unsafe { self.get_unchecked() })
>         } else {
>             None
>         }
>     }
> 
>     fn is_initialized(&self) -> bool {
>         self.once.is_completed() // <-- synchronization here!
>     }
> }
> ```
> 
> ```rust
> impl sys::Once {
>     pub fn is_completed(&self) -> bool {
>         // Use acquire ordering to make all initialization changes visible to the
>         // current thread.
>         self.state.load(Acquire) == COMPLETE
>     }
> }
> ```
> 

Yes it is not free, without `unwrap_unchecked()`. with `unwrap_unchecked()` however it does seem basically free (there is i think a line of assembly in difference, but it doesn't seem very important or relevant, and it's not a branch or anything expensive). because the compiler knows that if `self.is_initialized()` was `false` then there would be UB. so it can just remove the entire branch. that's why `unwrap_unchecked` exists for `Option`. it's to tell the compiler to just assume it can't be `None`. If we can't rely on this example to be optimized, then we can basically not rely on using `unwrap_unchecked()` at all since it'd very rarely be better. like `unwrap_unchecked()` internally uses the same pattern of having a branch where one of the arms leads to UB. and letting the compiler optimize the branch out https://doc.rust-lang.org/src/core/option.rs.html#1035 
```rs
pub const unsafe fn unwrap_unchecked(self) -> T {
        debug_assert!(self.is_some());
        match self {
            Some(val) => val,
            // SAFETY: the safety contract must be upheld by the caller.
            None => unsafe { hint::unreachable_unchecked() },
        }
    } 
```

why is it fine to trust that this branch gets optimized out and the call is free, but it's not fine to trust that the `.get().unwrap_unchecked()` branch gets optimized out? like both are just optimizations, there's no reason to assume one will get optimized and the other wont. if `.get().unwrap_unchecked()` is not free then neither should we count `.as_ref().unwrap_unchecked()` as free. 

another note btw, but on x86_64 at least, a `load(Acquire)` is actually free. 

and i have checked, if you use `unwrap_unchecked()`, then the compiler properly does understand that it doesn't need to check if the cell is initialized or not, and there is no branch emitted.

if `OnceLock` had a `get_unchecked`, then i dont think it would be any faster or better or even much clearer than just doing `.get().unwrap_unchecked()`.

Reply:
This is a quite advanced optimization though -- who is going to monitor that this behavior won't change and will be applicable in all possible platforms? I don't want to check the assembly for each export target that we eventually support.

Why put extra work on the LLVM optimizer if we know _a priori_ that a certain condition can't be true? Why use thread-safe locks in situation where we don't need synchronization? This is also about code clarity.

Let's use the simplest tool for the job, which is _guaranteed_ to not do extra work, and not dependent on quality-of-implementation.

---

I would like to conclude this discussion soon and focus on things that bring concrete benefit to the user. We can speculate endlessly about how all the cells interact, but in the end we have a simple problem to solve: **init once, read race-free**. If we can't find a simple solution using cells, then we need to go back to `static mut`.